modules/core/lib/Controller/Login.php

```diff
@@ -43,11 +64,391 @@ public function __construct(
+                UserPassBase::handleLogin($authStateId, $username, $password);
+            } catch (Error\Error $e) {
+                // Login failed. Extract error code and parameters, to display the error
+                $errorCode = $e->getErrorCode();
+                $errorParams = $e->getParameters();
+                $state['error'] = [
+                    'code' => $errorCode,
+                    'params' => $errorParams
+                ];
+                $authStateId = Auth\State::saveState($state, UserPassBase::STAGEID);
+                $queryParams = ['AuthState' => $authStateId];
+            }
+
+            if (isset($state['error'])) {
+                unset($state['error']);
+            }
+        }
+
+        $t = new Template($this->config, 'core:loginuserpass.twig');
+        $t->data['AuthState'] = $authStateId;
+
+        if (array_key_exists('forcedUsername', $state)) {
+            $t->data['username'] = $state['forcedUsername'];
+            $t->data['forceUsername'] = true;
+            $t->data['rememberUsernameEnabled'] = false;
+            $t->data['rememberUsernameChecked'] = false;
+            $t->data['rememberMeEnabled'] = $source->isRememberMeEnabled();
+            $t->data['rememberMeChecked'] = $source->isRememberMeChecked();
+        } else {
+            $t->data['username'] = $username;
+            $t->data['forceUsername'] = false;
+            $t->data['rememberUsernameEnabled'] = $source->getRememberUsernameEnabled();
+            $t->data['rememberUsernameChecked'] = $source->getRememberUsernameChecked();
+            $t->data['rememberMeEnabled'] = $source->isRememberMeEnabled();
+            $t->data['rememberMeChecked'] = $source->isRememberMeChecked();
+            if ($request->cookies->has($source->getAuthId() . '-username')) {
+                $t->data['rememberUsernameChecked'] = true;
+            }
+        }
+
+        $t->data['links'] = $source->getLoginLinks();
+        $t->data['errorcode'] = $errorCode;
+        $t->data['errorcodes'] = Error\ErrorCodes::getAllErrorCodeMessages();
+        $t->data['errorparams'] = $errorParams;
+        if (!empty($queryParams)) {
+            $t->data['queryParams'] = $queryParams;
+        }
+
+        if (isset($state['SPMetadata'])) {
+            $t->data['SPMetadata'] = $state['SPMetadata'];
+        } else {
+            $t->data['SPMetadata'] = null;
+        }
+
+        if ($cookie !== null) {
+            $t->headers->setCookie($cookie);
+        }
+
+        return $t;
+    }
+
+
+    /**
+     * This page shows a username/password/organization login form, and passes information from
+     * into the \SimpleSAML\Module\core\Auth\UserPassBase class, which is a generic class for
+     * username/password/organization authentication.
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @return \SimpleSAML\XHTML\Template
+     */
+    public function loginuserpassorg(Request $request): Template
```

Comment:
This is extremely duplicated from above.

Reply:
Maybe we can put some common code in private methods on the controller, or make or make one controller smart enough to deal with both cases.. They already share a template together..

Reply:
Yes, I think this is one reason why we'd do controllers. So we can get rid of duplicated processing easily. So if we move to controllers we should not be adding nearly the same code twice.

Reply:
So the current code also has all this duplication. So in a sense, merging this should be OK because it improves over the current situation, albeit not solving all its problems.

Getting rid of (more parts of) the duplication would of course make it even better.